flang/lib/Lower/OpenMP.cpp

```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+
+  if (blockDirective.v == llvm::omp::OMPD_parallel)
+    createParallelOp<Fortran::parser::OmpBeginBlockDirective>(
+        converter, eval,
+        std::get<Fortran::parser::OmpBeginBlockDirective>(blockConstruct.t));
+  else if (blockDirective.v == llvm::omp::OMPD_master) {
+    llvm::ArrayRef<mlir::Type> argTy;
+    auto &firOpBuilder = converter.getFirOpBuilder();
+    auto currentLocation = converter.getCurrentLocation();
     auto masterOp =
         firOpBuilder.create<mlir::omp::MasterOp>(currentLocation, argTy);
     createBodyOfOp<omp::MasterOp>(masterOp, firOpBuilder, currentLocation);
   }
 }
 
+static void genOMP(Fortran::lower::AbstractConverter &converter,
+                   Fortran::lower::pft::Evaluation &eval,
+                   const Fortran::parser::OpenMPLoopConstruct &loopConstruct) {
+
+  auto &firOpBuilder = converter.getFirOpBuilder();
+  auto currentLocation = converter.getCurrentLocation();
+  SmallVector<Value, 4> lowerBound, upperBound, step, privateClauseOperands,
+      firstPrivateClauseOperands, lastPrivateClauseOperands, linearVars,
+      linearStepVars;
+  mlir::Value scheduleChunkClauseOperand;
+  Attribute scheduleClauseOperand, collapseClauseOperand, noWaitClauseOperand,
+      orderedClauseOperand, orderClauseOperand;
+  const auto &wsLoopOpClauseList = std::get<Fortran::parser::OmpClauseList>(
+      std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t).t);
+  if (llvm::omp::OMPD_parallel_do ==
+      std::get<Fortran::parser::OmpLoopDirective>(
+          std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t).t)
+          .v)
+    createParallelOp<Fortran::parser::OmpBeginLoopDirective>(
+        converter, eval,
+        std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t));
+  else {
+    for (const auto &clause : wsLoopOpClauseList.v) {
+      if (const auto &privateClause =
+              std::get_if<Fortran::parser::OmpClause::Private>(&clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList = privateClause->v;
+        genObjectList(ompObjectList, converter, privateClauseOperands);
+      } else if (const auto &firstPrivateClause =
+                     std::get_if<Fortran::parser::OmpClause::Firstprivate>(
+                         &clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList =
+            firstPrivateClause->v;
+        genObjectList(ompObjectList, converter, firstPrivateClauseOperands);
+      }
+    }
+  }
+  for (const auto &clause : wsLoopOpClauseList.v) {
+    if (const auto &lastPrivateClause =
+            std::get_if<Fortran::parser::OmpClause::Lastprivate>(&clause.u)) {
+      const Fortran::parser::OmpObjectList &ompObjectList =
+          lastPrivateClause->v;
+      genObjectList(ompObjectList, converter, lastPrivateClauseOperands);
+    }
+  }
+  auto &doStmtEval = eval.getFirstNestedEvaluation();
+  auto *doStmt = doStmtEval.getIf<Fortran::parser::NonLabelDoStmt>();
```

Comment:
You might want to check if this succeed.

Reply:
Thanks for review @clementval 
Yes, I've another dev version which have this early check. This is the first draft, some clauses and tests are pending.
Waiting on the next Main to FIR rebase.